### ISSUE TITLE
fputsx is not round-trip safe with backslash-ending usernames

### ISSUE THREAD

**Issue by zeha** [shadow-maint/shadow#1055]
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1076619 reports:

The following sequence breaks /etc/group:

```
$ useradd 'vijai\'
$ usermod -aG sudo 'vijai\'
$ useradd blafasel
$ usermod -aG sudo blafasel
```

Results in:
```
sudo:x:27:install,vijaiaudio:x:29:install
```

Should have been:
```
sudo:x:27:install,vijai\,blafasel
audio:x:29:install
```

I believe this is caused by `fgetsx` "supporting" backslash as line-continuation indicator, but
* fputsx does not escape backslashes before line end
* there's a comment in fputsx indicating this is unsupported by everything else anyway.

**Comment by alejandro-colomar** [comment]
I had been looking at that code, and it seemed too brittle, and inconsistent.

I preferred to not change it at all while it worked, but your comment seems to be what I was waiting for: a report that it actually doesn't work.

Do you have any specific fix in mind?

Should we remove support for line continuations completely?  That would keep it simple, by calling standard fgets(3).  (But I was worried that it could break existing config files.)

**Comment by zeha** [comment]
> But I was worried that it could break existing config files.

From what I can tell, line continuation doesn't work on glibc systems for lookup:

```
root@tiksta:~# tail -n2 /etc/group
tftp:x:112:\
ch
root@tiksta:~# getent group tftp
tftp:x:112:\

root@tiksta:~# tail -n1 /etc/group
tftp:x:112:ch
root@tiksta:~# getent group tftp
tftp:x:112:ch
```

Thus I cannot imagine it being used a lot, and wouldn't worry about breaking config files (which were already broken if used).

> Do you have any specific fix in mind?

I was thinking replace fgetsx/fputsx by fgets/fputs.

**Comment by alejandro-colomar** [comment]
> > But I was worried that it could break existing config files.
> 
> From what I can tell, line continuation doesn't work on glibc systems for lookup:
> 
> ```
> root@tiksta:~# tail -n2 /etc/group
> tftp:x:112:\
> ch
> root@tiksta:~# getent group tftp
> tftp:x:112:\
> 
> root@tiksta:~# tail -n1 /etc/group
> tftp:x:112:ch
> root@tiksta:~# getent group tftp
> tftp:x:112:ch
> ```
> 
> Thus I cannot imagine it being used a lot, and wouldn't worry about breaking config files (which were already broken if used).
> 
> > Do you have any specific fix in mind?
> 
> I was thinking replace fgetsx/fputsx by fgets/fputs.

Sounds like what I had in mind.  I'll have a try.  :-)

**Comment by alejandro-colomar** [comment]
> https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1076619 reports:
> 
> The following sequence breaks /etc/group:
> 
> ```
> $ useradd 'vijai\'
> $ usermod -aG sudo 'vijai\'
> $ useradd blafasel
> $ usermod -aG sudo blafasel
> ```
> 
> Results in:
> 
> ```
> sudo:x:27:install,vijaiaudio:x:29:install
> ```
> 
> Should have been:
> 
> ```
> sudo:x:27:install,vijai\,blafasel
> audio:x:29:install
> ```
> 
> I believe this is caused by `fgetsx` "supporting" backslash as line-continuation indicator, but
> 
>     * fputsx does not escape backslashes before line end
> 
>     * there's a comment in fputsx indicating this is unsupported by everything else anyway.

On the other hand, usernames are not allowed to contain a `\`; do they?

<https://github.com/shadow-maint/shadow/blob/f3f501c81c1f5fd253bf0d243971cd6868a2d89e/lib/chkname.c#L60>

**Comment by zeha** [comment]
> On the other hand, user and group names are not allowed to contain a \; do they?

I think `--force-badname` will skip this check.

(As such, the reproducer above doesn't work as-is with upstream shadow. It works without `--force-badname` in Debian because we carry a patch to relax the user/group-name check - which I eventually want to get rid of.)

**Comment by alejandro-colomar** [comment]
> > On the other hand, user and group names are not allowed to contain a `\;` do they?
> 
> I think `--force-badname` will skip this check.
> 
> (As such, the reproducer above doesn't work as-is with upstream shadow. It works without `--force-badname` in Debian because we carry a patch to relax the user/group-name check - which I eventually want to get rid of.)

What's your opinion?  Is fgetsx()/fputsx() still broken?

**Comment by zeha** [comment]
> What's your opinion? Do you still think fgetsx()/fputsx() are broken?

Yes.

**Comment by alejandro-colomar** [comment]
> > What's your opinion? Do you still think fgetsx()/fputsx() are broken?
> 
> Yes.

I tend to agree;  `--badname` is a flag we support, and these functions break its behavior.

**Comment by Zugschlus** [comment]
Usernames containing backslashes at least used ot be a  rather common phenomenon in mixed Unix/Windows environments, with Unix user names being like DOMAIN\username. And breaking basic unix configuration files like /etc/group is a rather severe misbehavior.

**Comment by alejandro-colomar** [comment]
The title is "fputsx is not round-trip safe with **backslash-ending** usernames".  It doesn't say anything about **backslash-containing** usernames.

**Comment by hallyn** [comment]
Sorry, I'm going to try and find time in between meetings to do a bit more
testing and comment, but I'm worried patch 1 is a regression, and I'm trying
to decide on priorities for patch 3.  Using static variables may be out of
vogue, and aesthetically I prefer what you have, but in security critical
applications, you don't wnat to be doing any more malloc+free than you need to,
and the comment about freeing is just not quite right :)  (You are *responsible*
for freeing it, to avoid memory leak, not "free it as soon as you are done or
else something will break")

**Comment by alejandro-colomar** [comment]
> Sorry, I'm going to try and find time in between meetings to do a bit more testing and comment,

Sure, and feel free to defer this for after the release, if you want to review it with more calm.

> but I'm worried patch 1 is a regression,

Yes, it could regress some systems --if there are any systems that currently escape newlines in /etc/passwd and /etc/shadow--.  On the other hand, it fixes some bugs at the same time: <https://github.com/shadow-maint/shadow/issues/1055>.

It's a trade-off I'm willing to accept.

We'd have to make it visible in the release notes, of course.

> and I'm trying to decide on priorities for patch 3. Using static variables may be out of vogue, and aesthetically I prefer what you have, but in security critical applications, you don't wnat to be doing any more malloc+free than you need to,

Another point of view is that by freeing more often, we allow other allocations from our own code to succeed, whereas they could fail in case we'd hold too much unused heap memory.

> and the comment about freeing is just not quite right :) (You are _responsible_ for freeing it, to avoid memory leak, not "free it as soon as you are done or else something will break")

Yup, I meant that.  I'll reword `s/must/are responsible for/`.

**Comment by alejandro-colomar** [comment]
I've removed this PR from the list for this release.  Let's discuss it after the release.  That will allow merging more simple PRs.

**Comment by hallyn** [comment]
On Fri, Dec 05, 2025 at 10:58:15AM -0800, Alejandro Colomar wrote:
> alejandro-colomar left a comment (shadow-maint/shadow#1055)
> 
> I've removed this PR from the list for this release.  Let's discuss it after the release.  That will allow merging more simple PRs.

Actually, I'm good with this patch.

Note that the way the issue is presented is already a non-issue, as
   adduser 'user\'
will fail due to badname.  But indeed, the multiline support claimed
by fgetsx isn't providing anything useful.

Please feel free to re-mark it not draft, and we'll merge it.

**Comment by alejandro-colomar** [comment]
Oh, sure.  :)  Thanks!  Done!
